niceman/distributions/docker.py

```diff
@@ -0,0 +1,169 @@
+from .base import Distribution
+from .base import DistributionTracer
+from .base import TypedList
+from .base import _register_with_representer
+from ..dochelpers import borrowdoc
+from ..support.exceptions import CommandError
+
+
+@attr.s(slots=True, frozen=True, cmp=False, hash=True)
+class DockerImage(Package):
+    """Docker image information"""
+    id = attr.ib()
+    # Optional
+    architecture = attr.ib(default=None)
+    operating_system = attr.ib(default=None)
+    docker_version = attr.ib(default=None)
+    repo_digests = attr.ib(default=None)
+    repo_tags = attr.ib(default=None)
+    created = attr.ib(default=None)
+
+_register_with_representer(DockerImage)
+
+@attr.s
+class DockerDistribution(Distribution):
+    """
+    Class to provide commands to Docker engine.
+    """
+
+    images = TypedList(DockerImage)
+
+    def initiate(self, session):
+        """
+        Perform any initialization commands needed in the environment
+        environment.
+
+        Parameters
+        ----------
+        session : object
+            The Session to work in
+        """
+
+        # Raise niceman.support.exceptions.CommandError exception if Docker 
+        # engine is not to be found.
+        session.execute_command(['docker', 'info'])
+
+    def install_packages(self, session=None):
+        """
+        Install the Docker images associated to this distribution by the
+        provenance into the environment.
+
+        Parameters
+        ----------
+        session : object
+            Session to work in
+        """
+
+        for image in self.images:
+
+            # First, look in local Docker engine.
+            try:
+                session.execute_command(['docker', 'image', 'inspect',
+                    image.id])
+                continue
+            except CommandError:
+                pass
+
+            # Then, look in repos.
+            found = False
+            for digest in image.repo_digests:
+                try:
+                    session.execute_command(['docker', 'pull', digest])
+                    session.execute_command(['docker', 'tag', image.id,
+                        image.repo_tags[0]])
+                    found = True
+                    break
+                except CommandError:
+                    pass
+
+            # Can't find the image, so complain.
+            if not found:
+                raise CommandError(cmd='docker pull {}'.format(digest),
+                    msg="Unable to locate Docker image {}".format(image.id))
+
+_register_with_representer(DockerDistribution)
+
+
+class DockerTracer(DistributionTracer):
+    """Docker image tracer
+
+    If a Docker engine is not found running in the session, the files
+    are quietly passed on to thep next tracer.
+    """
+
+    HANDLES_DIRS = False
+
+    @borrowdoc(DistributionTracer)
+    def identify_distributions(self, files):
+        if not files:
+            return
+
+        images = []
+        remaining_files = []
```

Comment:
You end up converting this to a set at the end, so why not start off with a set?